runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,446 @@
+// BankForks::set_root()'s pruning.
+pub(crate) struct BankWithScheduler(Arc<Bank>);
+
+impl BankWithScheduler {
+    pub(crate) fn new(bank: Arc<Bank>) -> Self {
+        Self(bank)
+    }
+
+    pub(crate) fn bank_cloned(&self) -> Arc<Bank> {
+        self.0.clone()
+    }
+
+    pub(crate) fn bank(&self) -> &Arc<Bank> {
+        &self.0
+    }
+
+    pub(crate) fn into_bank(self) -> Arc<Bank> {
+        let bank = self.bank_cloned();
+        drop(self);
+        bank
+    }
+}
+
+impl Drop for BankWithScheduler {
+    fn drop(&mut self) {
+        self.0.schedule_termination();
+    }
+}
+
+impl Deref for BankWithScheduler {
+    type Target = Bank;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl BankForks {
+    pub fn install_scheduler_pool(&mut self, pool: SchedulerPoolArc) {
+        info!("Installed new scheduler_pool into bank_forks: {:?}", pool);
+        assert!(self.scheduler_pool.replace(pool).is_none());
+    }
+
+    pub(crate) fn install_scheduler_into_bank(&self, bank: &Arc<Bank>) {
+        if let Some(scheduler_pool) = &self.scheduler_pool {
+            let context = SchedulingContext::new(SchedulingMode::BlockVerification, bank.clone());
+            bank.install_scheduler(scheduler_pool.take_from_pool(context));
+        }
+    }
+}
+
+impl Bank {
+    pub fn install_scheduler(&self, scheduler: SchedulerBox) {
+        let mut scheduler_guard = self.scheduler.write().expect("not poisoned");
+        assert!(scheduler_guard.0.replace(scheduler).is_none());
+    }
+
+    pub fn with_scheduler(&self) -> bool {
+        self.scheduler.read().expect("not poisoned").0.is_some()
+    }
+
+    pub fn with_scheduling_context(&self) -> bool {
+        self.scheduler
+            .read()
+            .expect("not poisoned")
+            .0
+            .as_ref()
+            .and_then(|scheduler| scheduler.scheduling_context())
+            .is_some()
+    }
+
+    pub fn schedule_transaction_executions<'a>(
+        &self,
+        transactions: &[SanitizedTransaction],
+        transaction_indexes: impl Iterator<Item = &'a usize>,
+    ) {
+        trace!(
+            "schedule_transaction_executions(): {} txs",
+            transactions.len()
+        );
+
+        let scheduler_guard = self.scheduler.read().expect("not poisoned");
+        let scheduler = scheduler_guard.0.as_ref().expect("active scheduler");
+
+        for (sanitized_transaction, &index) in transactions.iter().zip(transaction_indexes) {
+            scheduler.schedule_execution(sanitized_transaction, index);
+        }
+    }
+
+    fn schedule_termination(&self) {
+        let mut scheduler_guard = self.scheduler.write().expect("not poisoned");
+        if let Some(scheduler) = scheduler_guard.0.as_mut() {
+            scheduler.schedule_termination();
+        }
+    }
+
+    #[must_use]
+    fn wait_for_scheduler(&self, reason: WaitReason) -> Option<ResultWithTimings> {
+        debug!(
```

Comment:
you meant like this?:

```rust
$ git diff HEAD~
diff --git a/runtime/src/installed_scheduler_pool.rs b/runtime/src/installed_scheduler_pool.rs
index 83408d1c861..c56c651f238 100644
--- a/runtime/src/installed_scheduler_pool.rs
+++ b/runtime/src/installed_scheduler_pool.rs
@@ -242,8 +242,7 @@ impl Bank {
     #[must_use]
     fn wait_for_scheduler(&self, reason: WaitReason) -> Option<ResultWithTimings> {
         debug!(
-            "wait_for_scheduler(slot: {}, reason: {reason:?}): started...",
-            self.slot()
+            "wait_for_scheduler(slot: {self.slot()}, reason: {reason:?}): started...",
         );
 
         let mut scheduler_guard = self.scheduler.write().expect("not poisoned");
```

actually, this is syntax error:

```
error: invalid format string: expected `'}'`, found `'.'`
   --> runtime/src/installed_scheduler_pool.rs:245:44
    |
245 |             "wait_for_scheduler(slot: {self.slot()}, reason: {reason:?}): started...",
    |                                       -    ^ expected `}` in format string
    |                                       |
    |                                       because of this opening brace
    |
    = note: if you intended to print `{`, you can escape it using `{{`

error: could not compile `solana-runtime` due to previous error

```

Reply:
ha I was unclear, but actually meant the opposite direction

```rust
         debug!(
            "wait_for_scheduler(slot: {}, reason: {:?}): started...",
            self.slot(),
            reason,
         );
```

It's fine to leave it, I always just find it harder to read if some params are inlined and others are not, and either choose to go all inlined or all not.

Reply:
I was also under the impression that it is recommended to use either all inlined, or all non-inlined arguments, but not a mix.
But I can not find any recommendations on this point right now.
One reason might be that there is a clippy rule that insists that all arguments should be inlined, when they all can be inlined.
But as soon as there is at least one non-inlinable argument, it stops complaining.
This rule has been [downgraded to "pedantic"](https://github.com/rust-lang/rust-clippy/issues/10087) due to `rust-analyzer` not been robust enough just yet.

Reply:
done: 1dc0c2cbf36322cefa48501c494d5e78b921cfe1